src/ratelimit_action.rs

```diff
@@ -0,0 +1,332 @@
+            .set(predicates)
+            .expect("Predicates must not be compiled yet!");
+
+        conditional_data
+            .data
+            .set(data_exp_list)
+            .expect("Data expressions must not be compiled yet!");
+        Ok(conditional_data)
+    }
+
+    fn predicates_apply(&self) -> bool {
+        let predicates = self
+            .predicates
+            .get()
+            .expect("predicates must be compiled by now");
+        predicates.is_empty()
+            || predicates.iter().all(|predicate| match predicate.test() {
+                Ok(b) => b,
+                Err(err) => {
+                    error!("Failed to evaluate {:?}: {}", predicates, err);
+                    panic!("Err out of this!")
+                }
+            })
+    }
+
+    pub fn entries(&self) -> RepeatedField<RateLimitDescriptor_Entry> {
+        if !self.predicates_apply() {
+            return RepeatedField::default();
+        }
+
+        let mut entries = RepeatedField::default();
+        for entry_builder in self
+            .data
+            .get()
+            .expect("data must be compiled by now")
+            .iter()
+        {
+            entries.push(entry_builder.evaluate());
+        }
+
+        entries
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct RateLimitAction {
+    grpc_service: Rc<GrpcService>,
+    scope: String,
+    service_name: String,
+    conditional_data_sets: Vec<ConditionalData>,
+}
+
+impl RateLimitAction {
+    pub fn new(action: &Action, service: &Service) -> Result<Self, String> {
+        Ok(Self {
+            grpc_service: Rc::new(GrpcService::new(Rc::new(service.clone()))),
+            scope: action.scope.clone(),
+            service_name: action.service.clone(),
+            conditional_data_sets: vec![ConditionalData::new(action)?],
+        })
+    }
+
+    pub fn build_descriptor(&self) -> RateLimitDescriptor {
+        let mut entries = RepeatedField::default();
+
+        for conditional_data in self.conditional_data_sets.iter() {
+            entries.extend(conditional_data.entries());
+        }
+
+        let mut res = RateLimitDescriptor::new();
+        res.set_entries(entries);
+        res
+    }
+
+    pub fn get_grpcservice(&self) -> Rc<GrpcService> {
+        Rc::clone(&self.grpc_service)
+    }
+
+    pub fn service(&self) -> &str {
+        self.service_name.as_str()
+    }
+
+    pub fn scope(&self) -> &str {
+        self.scope.as_str()
+    }
+
+    pub fn conditions_apply(&self) -> bool {
+        // For RateLimitAction conditions always apply.
+        // It is when building the descriptor that it may be empty because predicates do not
+        // evaluate to true.
+        true
+    }
+
+    pub fn get_failure_mode(&self) -> FailureMode {
+        self.grpc_service.get_failure_mode()
+    }
+
+    pub fn merge(&mut self, other: &RateLimitAction) {
+        self.conditional_data_sets
+            .extend_from_slice(&other.conditional_data_sets);
```

Comment:
that's why `GrpcService` needs to be `Clone`... here, on merge, it gets cloned and the `Rc` gets incremented. But... Yeah, well... if this works lets keep this for now.